pkg/backend/httpstate/client/client.go

```diff
@@ -519,3 +519,12 @@ func (pc *Client) AppendUpdateLogEntry(ctx context.Context, update UpdateIdentif
 	return pc.updateRESTCall(ctx, "POST", getUpdatePath(update, "log"), nil, req, nil,
 		updateAccessToken(token), httpCallOptions{RetryAllMethods: true})
 }
+
+// RecordEngineEvent posts an engine event to the Pulumi service.
+func (pc *Client) RecordEngineEvent(
+	ctx context.Context, update UpdateIdentifier, event apitype.EngineEvent, token string) error {
+	return pc.updateRESTCall(
```

Comment:
I was unaware the existing `log` endpoint was designed to handle rich events. But even then, I'd prefer to have a new endpoint that accepts a new, more specific type. This would just make it easier on the service-side to parse and ingest the engine event, since we could walk the specific event if we need to do something more complicated. (e.g. put it into different tables, update some index manually, etc.)

While the "union type" isn't great, it does provide an exact type definition for every event that we can accept.
```
type EngineEvent struct {
	CancelEvent      *CancelEvent       `json:"cancelEvent,omitempty"`
	StdoutEvent      *StdoutEngineEvent `json:"stdoutEvent,omitempty"`
	DiagnosticEvent  *DiagnosticEvent   `json:"diagnosticEvent,omitempty"`
	PreludeEvent     *PreludeEvent      `json:"preludeEvent,omitempty"`
	SummaryEvent     *SummaryEvent      `json:"summaryEvent,omitempty"`
	ResourcePreEvent *ResourcePreEvent  `json:"resourcePreEvent,omitempty"`
	ResOutputsEvent  *ResOutputsEvent   `json:"resOutputsEvent,omitempty"`
	ResOpFailedEvent *ResOpFailedEvent  `json:"resOpFailedEvent,omitempty"`
}
```

Whereas using the existing log handler, would need us to convert the `Fields` map to whatever specific type associated with the engine event, if we need it somewhere in the service API.

```
type AppendUpdateLogEntryRequest struct {
	Kind   string                 `json:"kind"`
	Fields map[string]interface{} `json:"fields"`
}
```

Is there some advantage to the logs endpoint I'm missing? Other than it can accept new engine events without requiring bumping the service dependency of pulumi/pulumi?